### PR TITLE
=str #22358 enable test and cleanup FutureFlatten source

### PR DESCRIPTION
**Pull request by ktoso**
Could not reproduce the leak in #22358
Likely due to new materializer being merged.

Fixed some things and added tests

### CODE REVIEW

**Comment by ktoso** [comment]
Please review @johanandren

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala [hunk 1]
```diff
@@ -139,30 +139,49 @@ class FutureFlattenSourceSpec extends StreamSpec {
...
     // Behaviour when inner source throws during materialization is undefined (leaks ActorGraphInterpreters)
     // until ticket #22358 has been fixed, this test fails because of it
-    "fail when the future source materialization fails" in pendingUntilFixed(assertAllStagesStopped {
-      val failure = TE("MatEx")
-
-      val (sourceMatVal, sinkMatVal) =
-        Source.fromFutureSource(
-          Future.successful(Source.fromGraph(new FailingMatGraphStage))
-        ).toMat(Sink.seq)(Keep.both)
+    "fail when the future source materialization fails" in assertAllStagesStopped {
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala]: remove the comment now that it works?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala [hunk 2]
```diff
@@ -139,30 +139,49 @@ class FutureFlattenSourceSpec extends StreamSpec {
...
+      subscriber.cancel()
+      publisher.expectCancellation()
+    }
+
     class FailingMatGraphStage extends GraphStageWithMaterializedValue[SourceShape[Int], String] {
       val out = Outlet[Int]("whatever")
       override val shape: SourceShape[Int] = SourceShape(out)
       override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, String) = {
-        throw TE("argh, materialization failed")
+        throw TE("INNER_FAILED")
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala]: Such a joykill!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala [hunk 3]
```diff
@@ -307,30 +311,30 @@ import scala.util.Try
...
-              val matValFuture = future.map { gr ⇒
-                val runnable = Source.fromGraph(gr).to(Sink.ignore)
-                interpreter.subFusingMaterializer.materialize(runnable, initialAttributes = attr)
-              }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+              val matValFuture = futureSource.map { gr ⇒
+                // downstream finish means it cancelled, so we push that signal through into the future materialized source
+                Source.fromGraph(gr).to(Sink.cancelled)
+                  .withAttributes(attr)
+                  .run()(subFusingMaterializer)
+              }(ExecutionContexts.sameThreadExecutionContext)
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: Ah, nice catch!

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala (hunk 2 above)
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala]: lol sorry (-:

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala (hunk 1 above)
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala]: ah yeah

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala (hunk 3 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: yup 👍

**Comment by ktoso** [comment]
Addressed review, ready to merge

**Review by jrudolph — APPROVED** [review]
LGTM

**Pull request merged**